### ISSUE TITLE
Implement torch loss reduction syntax

### ISSUE THREAD

**Issue by s-rog** [KevinMusgrave/pytorch-metric-learning#23]
In torch there is a `reduction='none'` parameter for losses and I would love to see it here as well. I skimmed over the code and it should be doable just bby modifying the base loss class?

I'm trying to implement my own miner as the implementation here doesn't seem to work with ddp. (Guessing some tensors aren't purged properly between epochs causing OOM at the start of epoch 2) But to do this I would need a non-reduced loss output.

Let me know what you think!

**Comment by KevinMusgrave** [comment]
@s-rog If reduction = 'none', should the loss return a loss per element in the batch, or per pair, or per triplet etc? 

Also re: ddp, not sure what issue you're having specifically, but maybe this issue is related: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/10#issuecomment-593170720

**Comment by s-rog** [comment]
Pytorch returns loss per element when you for example call `F.cross_entropy(x, y, reduction='none')`, but I haven't worked on pair-wise losses or triplet losses so not sure about those

**Comment by KevinMusgrave** [comment]
@s-rog I think returning the loss per element makes the most sense. Implementing this will be non-trivial though. Specifically, losses that operate on pairs and triplets will need to have a gather operation that sums up the loss for every anchor index. I'll try to implement this at the some point, but it might take a while.

**Comment by s-rog** [comment]
gotcha, thanks!

**Comment by KevinMusgrave** [comment]
I implemented this, though in a more flexible way. See the [reducers documentation](https://kevinmusgrave.github.io/pytorch-metric-learning/reducers/). If you want the raw output of the loss function, then use DoNothingReducer. The output of the loss function will then be a dictionary containing the losses and other info. I'll have to add documentation for what the raw output is exactly. 


Anyway, hopefully this extra level of modularity is helpful! It's available in v0.9.87

**Comment by KevinMusgrave** [comment]
Also let me know if you have suggestions etc. 😄
